File: tools/token_bench/programs/03_data_pipeline/data_pipeline.py
def lcg(seed):
    return (seed * 1103515245 + 12345) % 2147483648

records = []
seed = 42
for i in range(5000):
    seed = lcg(seed)
    score = seed % 100
    cat = score % 4
    category = {0: "alpha", 1: "beta", 2: "gamma"}.get(cat, "delta")
    records.append({"name": f"item{i}", "score": score, "category": category})

high_scores = [r for r in records if r["score"] >= 50]

transformed = [
    {"name": r["name"], "score": r["score"] * 2, "category": r["category"]}
    for r in high_scores
]

groups = {}
for r in transformed:
    key = r["category"]
    if key not in groups:
        groups[key] = []
    groups[key].append(r["score"])

for k, scores in groups.items():
    scores.sort()
    avg = sum(scores) // len(scores)
    print(f"{k}: count={len(scores)} avg={avg} min={scores[0]} max={scores[-1]}")

File: tools/token_bench/programs/03_data_pipeline/data_pipeline.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Record {
    public string Name;
    public int Score;
    public string Category;
}

class Program {
    static int Lcg(int seed) {
        return (int)(((long)seed * 1103515245 + 12345) % 2147483648);
    }

    static void Main() {
        var records = new List<Record>();
        int seed = 42;
        for (int i = 0; i < 5000; i++) {
            seed = Lcg(seed);
            int score = seed % 100;
            int cat = score % 4;
            string category = cat switch {
                0 => "alpha",
                1 => "beta",
                2 => "gamma",
                _ => "delta"
            };
            records.Add(new Record {Name = $"item{i}", Score = score, Category = category});
        }

        var highScores = records.Where(r => r.Score >= 50).ToList();

        var transformed = highScores.Select(r => new Record {
            Name = r.Name, Score = r.Score * 2, Category = r.Category
        }).ToList();

        var groups = new Dictionary<string, List<int>>();
        foreach (var r in transformed) {
            if (!groups.ContainsKey(r.Category))
                groups[r.Category] = new List<int>();
            groups[r.Category].Add(r.Score);
        }

        foreach (var (k, scores) in groups) {
            scores.Sort();
            int avg = scores.Sum() / scores.Count;
            Console.WriteLine($"{k}: count={scores.Count} avg={avg} min={scores[0]} max={scores[^1]}");
        }
    }
}
